deepcell/training/model_trainer.py

```diff
@@ -0,0 +1,486 @@
+            deepcell version
+            dataset
+            preprocessing function(s)
+            image generators
+            model (along with initialization state?)
+            postprocessing function(s)
+        """
+        # initialize metadata
+        self.output_metadata = {}
+        # store deepcell version
+        self.output_metadata["deepcell_version"] = __version__
+        # store dataset information
+        self.output_metadata["dataset"] = {}
+        self.output_metadata["dataset"]["training"] = {}
+        self.output_metadata["dataset"]["training"]["X_train_shape"] = self.X_train.shape
+        self.output_metadata["dataset"]["training"]["X_train_datatype"] = self.X_train.dtype.name
+        self.output_metadata["dataset"]["training"]["X_train_md5_digest"] = hashlib.md5(self.X_train).hexdigest()
+        self.output_metadata["dataset"]["training"]["y_train_shape"] = self.y_train.shape
+        self.output_metadata["dataset"]["training"]["y_train_datatype"] = self.y_train.dtype.name
+        self.output_metadata["dataset"]["training"]["y_train_md5_digest"] = hashlib.md5(self.y_train).hexdigest()
+        self.output_metadata["dataset"]["testing"] = {}
+        self.output_metadata["dataset"]["testing"]["X_test_shape"] = self.X_test.shape
+        self.output_metadata["dataset"]["testing"]["X_test_datatype"] = self.X_test.dtype.name
+        self.output_metadata["dataset"]["testing"]["X_test_md5_digest"] = hashlib.md5(self.X_test).hexdigest()
+        self.output_metadata["dataset"]["testing"]["y_test_shape"] = self.y_test.shape
+        self.output_metadata["dataset"]["testing"]["y_test_datatype"] = self.y_test.dtype.name
+        self.output_metadata["dataset"]["testing"]["y_test_md5_digest"] = hashlib.md5(self.y_test).hexdigest()
+
+    def _data_prep(self):
+        
+        def prep_generator(
+                generator_type,
+                output_metadata,
+                data_generator,
+                input_data_dict,
+                skip = None,
+                seed = None,
+                batch_size = None,
+                transform = None,
+                transform_kwargs = {}):
+            """
+            Get training and validation data back from data_generator.flow() and document every step of the process
+            using output_metadata.
+            """
+
+
+            generator_type_string = generator_type + "_generator"
+
+            output_metadata["generators"][generator_type_string] = {}
+            data_generator_name = re.match("<([\w.]+) object at",str(data_generator)).group(1)
+            output_metadata["generators"][generator_type_string]["class"] = data_generator_name
+            
+            parameters = {}
+            parameters["input_data_dict"] = {}
+            for entry in input_data_dict:
+                parameters["input_data_dict"][entry] = {}
+                parameters["input_data_dict"][entry]["md5_digest"] = hashlib.md5(input_data_dict[entry]).hexdigest()
```

Comment:
I think a helper function `_get_hash` or something would be nice to guarantee that it's always being hashed with the same algo. Would be eaiser to change the algo that way too if we ever wanted to do so.